BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -115,22 +115,45 @@ def new_personal #:nodoc:
     @provider         = UserkeyFlatDirSshDataProvider.new( :user_id   => current_user.id,
                                                            :group_id  => provider_group_id,
                                                            :online    => true,
-                                                           :read_only => false
+                                                           :read_only => false,
                                                          )
     @groups           = current_user.assignable_groups
   end
 
-  # create by normal user,  only UserkeyFlatDirSshDataProvider
+  # Create by normal user, only UserkeyFlatDirSshDataProvider
+  # S3FlatDataProvider, S3MultiLevelDataProvider
   def create_personal
-    normal_params = params.require_as_params(:data_provider)
-                          .permit(:name, :description, :group_id,
-                                  :remote_user, :remote_host,
-                                  :remote_port, :remote_dir
-                                 )
+    # Check if the type is an authorized class
+    dp_type         = (params[:type] || "UserkeyFlatDirSshDataProvider").constantize
+    authorized_type = [UserkeyFlatDirSshDataProvider, S3FlatDataProvider, S3MultiLevelDataProvider]
+    if ! authorized_type.include?(dp_type)
+      flash[:error] = "DataProvider type is not authorized"
+      respond_to do |format|
+        format.html { render :action => :new_personal}
+        format.json { render :json   => { :error => "DataProvider type is not authorized" },  :status => :unprocessable_entity }
+      end
+    end
+
+    # Whitelist params
+    whitelist_params = dp_type == UserkeyFlatDirSshDataProvider ?
```

Comment:
Rewrite the next 12 lines to call a separete method to white list the params, one for UserKey DPs and one for S3 DPs.